File: app/src/main/java/com/rafaelmallare/terraphage/Character.java
package com.rafaelmallare.terraphage;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import static com.rafaelmallare.terraphage.AttributeType.ARM;
import static com.rafaelmallare.terraphage.AttributeType.ATK;
import static com.rafaelmallare.terraphage.AttributeType.DEF;
import static com.rafaelmallare.terraphage.AttributeType.HP;
import static com.rafaelmallare.terraphage.AttributeType.INIT;
import static com.rafaelmallare.terraphage.AttributeType.MDMG;
import static com.rafaelmallare.terraphage.AttributeType.RDMG;
import static com.rafaelmallare.terraphage.AttributeType.REGEN;
import static com.rafaelmallare.terraphage.AttributeType.SPD;
import static com.rafaelmallare.terraphage.StatType.CHR;
import static com.rafaelmallare.terraphage.StatType.CON;
import static com.rafaelmallare.terraphage.StatType.DEX;
import static com.rafaelmallare.terraphage.StatType.INT;
import static com.rafaelmallare.terraphage.StatType.PER;
import static com.rafaelmallare.terraphage.StatType.STR;

/**
 * Created by Rj on 10/19/2016.
 */
public class Character {
    private static Character ourInstance = new Character();

    public static Character getInstance() {
        return ourInstance;
    }

    private Character() {
        mUnarmed = new Weapon("Unarmed", 0, WeaponType.Melee);

        mName = "Default Name";
        mHomeland = "Default Homeland";
        mSil = 0;
        mExp = 0;

        mAnima = 0;

        mEquippedWeapon = null;
        mInventoryList = new ArrayList<>();
        mInventoryList.add(mUnarmed);

        mRefID = 1;

        mPhysicalStats = new StatGroup(CON, STR, DEX);
        mMentalStats = new StatGroup(PER, CHR, INT);

        mAttributeList = new ArrayList<>();
        mAttributeList.add(new Attribute(HP));
        mAttributeList.add(new Attribute(REGEN));
        mAttributeList.add(new Attribute(INIT));
        mAttributeList.add(new Attribute(SPD));
        mAttributeList.add(new Attribute(DEF));
        mAttributeList.add(new Attribute(ATK));
        mAttributeList.add(new Attribute(MDMG));
        mAttributeList.add(new Attribute(RDMG));
        mAttributeList.add(new Attribute(ARM));

        equipWeapon(mUnarmed);

    }

    private String mName;
    private String mHomeland;
    private int mSil;
    private int mExp;

    private int mAnima;

    private Weapon mEquippedWeapon;
    private Weapon mUnarmed;

    private StatGroup mPhysicalStats;
    private StatGroup mMentalStats;
    private ArrayList<Attribute> mAttributeList;

    private ArrayList<Gear> mInventoryList;
    private int mRefID;

    //region -REGION: Standard getters and setters-
    public String getName() {
        return mName;
    }

    public void setName(String name) {
        this.mName = mName;
    }

    public String getHomeland() {
        return mHomeland;
    }

    public void setHomeland(String Homeland) {
        this.mHomeland = mHomeland;
    }

    public int getSil() {
        return mSil;
    }

    public int getExp() {
        return mExp;
    }

    public void setExp(int mExp) {
        this.mExp = mExp;
    }

    public void setSil(int Sil) {
        this.mSil = mSil;
    }

    public int getAnima() {
        return mAnima;
    }

    public void setAnima(int Anima) {
        this.mAnima = mAnima;
    }

    public Weapon getEquippedWeapon(){
        return mEquippedWeapon;
    }

    public ArrayList<Gear> getInventory(){
        return mInventoryList;
    }

    public ArrayList<Weapon> getWeapons(){
        ArrayList<Weapon> weaponList = new ArrayList<>();
        for (int i = 0; i < mInventoryList.size(); i++){
            Gear gear = mInventoryList.get(i);
            if (gear instanceof Weapon){
                Weapon weapon = (Weapon) gear;
                weaponList.add(weapon);
            }
        }

        return weaponList;
    }
    //endregion

    public int getStat(StatType stat){
        int statVal = 0;

        if(stat.getCategory() == "PHYS"){
            statVal = mPhysicalStats.getStat(stat);
        } else if(stat.getCategory() == "MENT"){
            statVal = mMentalStats.getStat(stat);
        }

        return statVal;
    }

    public boolean increaseStat(StatType stat, int increaseBy){
        boolean successful = false;

        if(stat.getCategory() == "PHYS"){
            successful = mPhysicalStats.increaseStat(stat, increaseBy);
        } else if (stat.getCategory() == "MENT") {
            successful = mMentalStats.increaseStat(stat, increaseBy);
        }

        return successful;
    }

    public Attribute getAttribute(AttributeType attributeType){
        return mAttributeList.get(attributeType.value());
    }

    public void addGear(Gear gear){
        Gear newGear = gear;
        newGear.setRefID(mRefID);
        mRefID++;

        mInventoryList.add(newGear);
    }

    public boolean removeGear(Gear gear){
        boolean success = false;
        if(gear.getRefID() != 0) {
            mInventoryList.remove(gear);
            success = true;

            if(mInventoryList.size() == 0){
                mRefID = 1;
            }
        }
        return success;
    }

    public void equipWeapon(Weapon weapon){
        mEquippedWeapon = weapon;
        weapon.equip();
        for(Attribute attribute : mAttributeList){
            attribute.addModifier(weapon);
        }
    }

    public void unequipWeapon(Weapon weapon){
        mEquippedWeapon = mUnarmed;
        weapon.unEquip();
        for(Attribute attribute : mAttributeList){
            attribute.removeModifier(weapon);
        }
    }

    /**********Attribute Class**********/
    public class Attribute {
        public Attribute(AttributeType attributeType){
            mType = attributeType;
            mBaseValue = 0;
            mTotalValue = 0;
            mModifiers = new HashMap<>();
        }

        public Attribute(AttributeType attributeType, int baseValue){
            mType = attributeType;
            mBaseValue = baseValue;
            mTotalValue = mBaseValue;
            mModifiers = new HashMap<>();
        }

        private AttributeType mType;
        private int mBaseValue;
        private int mTotalValue;
        private HashMap<String, Integer> mModifiers;

        public int getBaseValue(){
            return updateBaseValue();
        }

        public int getTotalValue(){
            return updateTotalValue();
        }

        public HashMap<String, Integer> getModifiers(){
            return mModifiers;
        }

        private int updateTotalValue(){
            mTotalValue = updateBaseValue();

            for(Map.Entry<String, Integer> modifier : mModifiers.entrySet()){
                mTotalValue += modifier.getValue();
            }

            return mTotalValue;
        }

        public int updateBaseValue(){
            switch(mType){
                case HP:    mBaseValue = mPhysicalStats.getStat(CON) + mPhysicalStats.getStat(STR) + 10;
                    break;
                case REGEN: mBaseValue = mPhysicalStats.getStat(CON) / 2;
                    break;
                case INIT:  mBaseValue = mPhysicalStats.getStat(DEX) + mMentalStats.getStat(PER);
                    break;
                case SPD:   mBaseValue = (mPhysicalStats.getStat(DEX) / 2) + 3;
                    break;

                case DEF:   mBaseValue = mPhysicalStats.getStat(DEX) /*+ Evasion skill*/;
                    break;
                case ATK:   mBaseValue = mMentalStats.getStat(PER) /*+ Weapon skill*/;
                    break;
                case MDMG:  mBaseValue = mPhysicalStats.getStat(STR);
                    break;
                case RDMG:  //No action - RDMG depends only on ranged weapon damage
                case ARM:   //No action - ARM depends only on gear ARM atttribute
                default:    break;
            }

            return mBaseValue;
        }

        public void addModifier(Gear gear){
            for(Map.Entry<AttributeType, Integer> modifier : gear.getModifierMap().entrySet()){
                if(modifier.getKey() == mType){
                    mModifiers.put(gear.getName(), modifier.getValue());
                }
            }

            updateTotalValue();
        }

        public void removeModifier(Gear gear){
            String gearName = gear.getName();
            if(mModifiers.containsKey(gearName)){
                mModifiers.remove(gearName);
            }

            updateTotalValue();
        }
    }

    /**********StatGroup Class**********/
    public class StatGroup {
        public StatGroup(StatType statOne, StatType statTwo, StatType statThree){
            mStatMap = new HashMap<>();

            mStatMap.put(statOne, 0);
            mStatMap.put(statTwo, 0);
            mStatMap.put(statThree, 0);
        }

        private HashMap<StatType, Integer> mStatMap;

        private boolean checkMaxDifference(StatType stat, int newValue){
            HashMap<StatType, Integer> tmpMap = new HashMap<>();
            tmpMap.putAll(mStatMap);

            tmpMap.put(stat, newValue);

            int max = 0;
            int min = 1000;
            for(Map.Entry<StatType, Integer> entry : tmpMap.entrySet()){
                int currentVal = entry.getValue();
                if(max < currentVal){
                    max = currentVal;
                }
                if(min > currentVal){
                    min = currentVal;
                }
            }

            if((max - min) > 4){
                return false;
            } else{
                return true;
            }
        }

        public int getStat(StatType stat){
            return mStatMap.get(stat);
        }

        public HashMap<StatType, Integer> getAllStats(){
            return mStatMap;
        }

        public boolean increaseStat(StatType stat, int increaseBy){
            int newValue = mStatMap.get(stat) + increaseBy;

            if(checkMaxDifference(stat, newValue)) {
                mStatMap.put(stat, newValue);
                return true;
            } else{
                return false;
            }
        }
    }
}

File: app/src/main/java/com/rafaelmallare/terraphage/Character.kt
package com.rafaelmallare.terraphage

import com.rafaelmallare.terraphage.AttributeType.*
import com.rafaelmallare.terraphage.StatType.*
import com.rafaelmallare.terraphage.WeaponType.*
import com.rafaelmallare.terraphage.GearType.*
import kotlin.coroutines.experimental.EmptyCoroutineContext.plus

/**
 * Created by Rj on 5/23/2017.
 */
object Character {
    var name = "Default Name"
    var homeland = "Default Homeland"
    var sil = 0

    var expTotal = 0
        set(value) {
            if (value > field) {
                expCurrent += value - field
                field = value
            }
        }

    var expCurrent = expTotal

    var anima = 0

    object physStats : StatGroup(CON, STR, DEX)
    object mentStats : StatGroup(PER, CHR, INT)

    val skills = Skills

    val attributes = listOf(Attribute(HP), Attribute(REGEN), Attribute(INIT), Attribute(SPD),
                            Attribute(DEF), Attribute(ATK), Attribute(MDMG), Attribute(RDMG),
                            Attribute(ARM))

    private var maxHealth = getAttribute(HP)
    var currentHealth = maxHealth
        set(value) {
            if (value > maxHealth) field = maxHealth
            else if (value < 0) field = 0
            else field = value
        }

    val unarmed = Gear("Unarmed", Weapon, Melee, 0)
    val unarmored = Gear("Unarmored", Armor, null, 0)
    var equippedWeapon = unarmed
    var equippedArmor = unarmored
    val inventory = mutableListOf<Gear>()
    val weaponInv: List<Gear>
        get() = inventory.filter { gear -> gear.type == Weapon }
    val armorInv: List<Gear>
        get() = inventory.filter { gear -> gear.type == Armor }
    val gadgetInv: List<Gear>
        get() = inventory.filter { gear -> gear.type == Gadget }

    fun getStat(stat: StatType) : Int = if (stat == ANI) anima else if (physStats.contains(stat)) physStats.get(stat) else mentStats.get(stat)

    fun increaseStatBy(stat: StatType, value: Int) {
        if (physStats.contains(stat)) physStats.increaseStatBy(stat, value)
        else if (mentStats.contains(stat)) mentStats.increaseStatBy(stat, value)
        else anima += value
    }

    fun getAttribute(attribute: AttributeType) = attributes[attribute.ordinal].totalValue

    fun equipWeapon(weapon: Gear) {
        if (equippedWeapon != unarmed) unequipWeapon(equippedWeapon)
        if (weapon in weaponInv) equippedWeapon = weapon
        for (attribute in attributes) attribute.addModifier(weapon)
    }

    fun unequipWeapon(weapon: Gear) {
        if (weapon == equippedWeapon) equippedWeapon = unarmed
        for (attribute in attributes) attribute.removeModifier(weapon)
    }

    fun equipArmor(armor: Gear) {
        if (equippedArmor != unarmored) unequipArmor(equippedArmor)
        if (armor in armorInv) equippedArmor = armor
        for (attribute in attributes) attribute.addModifier(armor)
    }

    fun unequipArmor(armor: Gear) {
        if (armor == equippedArmor) equippedArmor = unarmored
        for (attribute in attributes) attribute.removeModifier(armor)
    }

    fun increaseSkill(skill: String) : Boolean {
        if ((expCurrent - 5) >= 0) {
            skills[skill] = skills[skill] + 1
            expCurrent -= 5
            return true
        } else {
            return false
        }
    }

    class Attribute(val type: AttributeType) {
        val modifiers = mutableMapOf<String, Int>()

        val baseValue: Int
            get() {
                val value: Int
                when (type) {
                    AttributeType.HP -> {
                        value = getStat(CON) + physStats.get(STR) + 10
                        maxHealth = value
                        currentHealth = maxHealth
                    }
                    REGEN -> value = getStat(CON) / 2
                    INIT -> value = getStat(DEX) + mentStats.get(PER)
                    SPD -> value = (getStat(DEX) / 2) + 3
                    DEF -> value = getStat(DEX) + skills["Evasion", true].toInt()
                    ATK -> value = getStat(PER) + skills["Weapon Mastery", true].toInt()
                    MDMG -> value = getStat(STR)
                    RDMG -> value = if (equippedWeapon.subType == Ranged) (equippedWeapon.modifiers[RDMG] ?: 0) else 0
                    ARM -> value = (equippedArmor.modifiers[ARM] ?: 0) + (equippedWeapon.modifiers[ARM] ?: 0)
                    else -> value = 0
                }

                return value
            }

        val totalValue: Int
            get() {
                var value = baseValue

                if (type != RDMG && type != ARM) {
                    for (item in modifiers.values) {
                        value += item
                    }
                }

                return value
            }

        fun addModifier(gear: Gear) {
            if (gear.modifiers.containsKey(type)) modifiers.put(gear.name, gear.modifiers[type] ?: 0)
        }

        fun removeModifier(gear: Gear) {
            if (modifiers.containsKey(gear.name)) modifiers.remove(gear.name)
        }
    }

    open class StatGroup(statOne: StatType, statTwo: StatType, statThree: StatType) {
        val statMap = mutableMapOf(Pair(statOne, 0), Pair(statTwo, 0), Pair(statThree, 0))

        private fun withinConstraint(stat: StatType, value: Int): Boolean{
            val tmpMap = statMap.toMutableMap()
            tmpMap.put(stat, (tmpMap[stat] ?: 0) + value)

            return ((tmpMap.values.max() ?: 0) - (tmpMap.values.min() ?: 0)) <= 4
        }

        fun increaseStatBy(stat: StatType, value: Int) {
            if (withinConstraint(stat, value)) {
                statMap.put(stat, (statMap[stat] ?: 0) + value)
            }
        }

        fun get(stat: StatType) : Int {
            return statMap[stat] ?: -1
        }

        fun contains(stat: StatType) : Boolean {
            return statMap.contains(stat)
        }
    }
}